[WIP] Initial structure

kevinyamauchi opened the pull request:
This is an initial proposal for the structure for the library and based on discussion with @haesleinhuepf . 

The core of analysis system is a registry of functions that take one image type and return a pandas dataframe. For measurements on label images, the input type is LabelImage and the measurement function should return a dataframe where the index is the label value for each measurement.

The types are defined in `morphometrics.types` and include:
```python
# pixel values are signal magnitude
IntensityImage = np.ndarray

# pixel values are a calculated feature value
FeatureImage = np.ndarray

# pixel values are bool
BinaryImage = np.ndarray

# pixel values are class/instance membership
LabelImage = np.ndarray

# must index must be the label value
LabelMeasurementTable = pd.DataFrame
```
Analysis functions are registered with a decorator. For example:

```python
@register_measurement(name="volume")
def volume(label_im: LabelImage) -> LabelMeasurementTable:
    rp_table = regionprops_table(label_im, properties=("label", "area"))

    return pd.DataFrame(rp_table).set_index("label")
```

See the example `registered_measurements.py` to see how one checks what measurements are available and then applies them to an image.

```python
from morphometrics.data import simple_labeled_cube
from morphometrics.measure import available_measurments, measure_all

print(f"available_measurements: {available_measurments()}")
>>> available_measurements: ['volume', 'centroid']

label_im = simple_labeled_cube()

measurement_table = measure_all(label_im)
print(measurement_table)

>>> 
       area  centroid-0  centroid-1  centroid-2
label                                          
1       125         7.0         7.0         7.0
2       125         7.0         2.0         2.0
3       500         2.0         4.5         4.5
```

Next step is to add a napari widget plugin to view the results of the measurement in a table.

kevinyamauchi commented:
Some initial thoughts:

- In general I like the registration system. I think it can be flexible and allows users to make their custom functions available. The challenge is of course name space, but I think we can at least add some stuff to check for collisions and warn before clobbering (or modify names).
- I'd like to add some specificity to the function registries. For example, I could imagine separate registries for measurements on label images and meshes (they have different function signatures), so we should bake that into the initial names (e.g., `_measurements` -> `_label_measurements`)
- we should probably add the ability to pass both a LabelImage and a FeatureImage or IntensityImage for quantification of features in masked areas or instances (e.g., mean intensity)

haesleinhuepf commented on setup.cfg:
```diff
@@ -1,26 +1,28 @@
-    Operating System :: OS Independent
-    License :: OSI Approved :: BSD License
+    Programming Language :: Python :: 3.10
+    Topic :: Software Development :: Testing
```
```suggestion
    Topic :: Scientific/Engineering :: Image Processing
```

haesleinhuepf commented on src/morphometrics/measure/label.py:
```diff
@@ -0,0 +1,19 @@
+
+@register_measurement(name="volume")
+def volume(label_im: LabelImage) -> LabelMeasurementTable:
+    rp_table = regionprops_table(label_im, properties=("label", "area"))
+
+    return pd.DataFrame(rp_table).set_index("label")
+
+
+@register_measurement(name="centroid")
+def centroid(label_im: LabelImage) -> LabelMeasurementTable:
+    rp_table = regionprops_table(label_im, properties=("label", "centroid"))
```
That's a tricky pattern. If I'm not wrong, this will have sub-optimal performance, especially on big images. We might want to explore a pattern where we can cache the `regionprops` object.

haesleinhuepf replied:
Let me just add a link. I used an alternative pattern, that could be a starting point. Imagine there is an object that holds `regionprops`, in case any function in a pool of functions requested it. If a second functions asks for `regionprops`, it would access the same object as the first function used before. Not sure if the term is right, but that's some kind of "singleton" on `regionprops`-object level. You can read an implementation of this [here](https://github.com/clij/clijx-accelerated-pixel-and-object-classification/blob/main/src/main/java/net/haesleinhuepf/clijx/apoc/LabelFeatureGenerator.java#L221) (sorry, in Java)

kevinyamauchi replied:
Agreed. As discussed below, I've introduced the concept of "measurement sets" with selectable measurements to allow groups of measurements be easily computed with cacheing.

haesleinhuepf reviewed (approved):
Hi Kevin @kevinyamauchi ,

this looks super cool! Just two remarks and a question:
* I think we might need an approach to cache regionprops objects and similar (in simple-itk, there is a [similar mechanism](https://github.com/haesleinhuepf/napari-simpleitk-image-processing/blob/main/src/napari_simpleitk_image_processing/_simpleitk_image_processing.py#L808)). See my comment in line. 
* I come from Java island, where we have long human readable variable names which allows reading out code loudly like human readable text. And I believe that's a good habit. Speaking about `label_image` on the phone has a high success rate in transfering information compared to a variable named `label_im`. I'm suggesting to rename all variables which end with '_im' to '_image'. Sorry for being picky ;-)

And the question: Do I understand right, this python library will serve as central point to collect measurements, similar to [this earlier approach](https://github.com/0x00b1/napari-features). I'm actually wondering: If I developed a super cool algorithm, e.g. to measure `circularity`, shall I `from morphometrics import register_measurement`? Or shall I send a pull-request with my `determine_circularity()` code to this repository as PR?

Best,
Robert

kevinyamauchi commented:
@haesleinhuepf , I've been thinking a bit more about the API and I've updated the PR. I am still unsure where to put this (e.g., new regionprops plugin, napari-skimage-regionprops, leave it here), but it was just easier to prototype here. All of that is to say I would like to continue our discussion on where the "core" region props measurement code should go separately and I would also find it helpful to discuss the implementation in this PR.

### measurement sets
I completely agree with your statement above that we should allow batches of measurements to be computed since this allows for caching etc.  Thus, I have introduced a new type of function that can be registered: "measurement sets". A measurement set is a function that takes a label image, an intensity image, and a set of kwargs for turning individual measurements on/off. I have modeled this after your `regionprops` function in `napari-skimage-regionprops` where each individual measurement is turned on/off by settings `measurement_name` to True/False in the kwargs. Like the single measurement function, the measurement set function must return a pandas DataFrame where the columns are the measurements and the index is the label value.


### GUI

I have made a prototype of dock widget plugin autogenerates a widget for each registered measurement. The individual measurements get a checkbox and the measurement sets get a collapsible section containing a checkbox for each measurement choice.

### additional kwargs
I think it would be fairly straightforward to extend this a bit more to allow additional kwargs to be passed to measurement functions. One example would be passing the filter parameters if one was summing the intensities in an object after filtering. I think we could use magicgui to generate the widgets for the GUI.


### models
I would also like to create small data classes or pydantic models to formalize parts of the  API (e.g., the function name and choices for the measurement set functions). This also will allow us to validate the options before running the measurements. This will be especially useful for measurements that take a long time to run as it will prevent the analysis from failing halfway through a long run due to invalid settings.

### demo

Here is how it all looks together. The demo images are available as sample data. Notice individual functions (`volume` and `centroid`) are displayed as individual checkboxes and your `regionprops` is in a collapsible widget with a checkbox for each measurement option. This stores the results in layer properties and displays them using your table widget (imported from `napari-skimage-regionprops`. Currently, the measurement is blocking, but we can put it into a separate thread and update a progress bar when each measurement completes.

https://user-images.githubusercontent.com/1120672/154799281-7ef657fe-5ad6-4783-8116-c2467699cb58.mov

kevinyamauchi commented:
> I come from Java island, where we have long human readable variable names which allows reading out code loudly like human readable text. And I believe that's a good habit. Speaking about label_image on the phone has a high success rate in transfering information compared to a variable named label_im. I'm suggesting to rename all variables which end with '_im' to '_image'. Sorry for being picky ;-)

Yes, I think this is a good point. I've updated the variable names to be complete words.

haesleinhuepf commented:
Hey Kevin,

great to see this emerging!

> am still unsure where to put this

I think it's partially overlapping with napari-features. Maybe, we can take it over and collect the best of both approaches there. See also: https://github.com/0x00b1/napari-features/issues/18

I'm a bit busy these days, but hope to join forces on this with you in March 🌞

kevinyamauchi commented:
Hey @haesleinhuepf ! Thanks for dropping by even thought you're super busy.

I agree there is overlap with `napari-features`. I am happy to combine this code over there if @0x00b1 is open to it. Thank you for starting that conversation.

I think I will go ahead and merge this PR for now so I can keep experimenting with other things. I do not mind moving/refactoring when we figure out the best path forward.

Looking forward to when you have more time in March!
